clients/hadoopfs/src/main/java/io/lakefs/MetadataClient.java

```diff
@@ -0,0 +1,68 @@
+package io.lakefs;
+
+import com.amazonaws.services.s3.model.GetObjectMetadataRequest;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import org.apache.commons.lang.NullArgumentException;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URI;
+
+public class MetadataClient {
+    public static final Logger LOG = LoggerFactory.getLogger(MetadataClient.class);
+    private final FileSystem fs;
+
+    public MetadataClient(FileSystem fs) {
+        if (fs == null) {
+            throw new NullArgumentException("fs");
+        }
+        this.fs = fs;
+    }
+
+    /**
+     * get object metadata by physical address
+     * @param physicalUri physical uri of object
+     * @return ObjectMetadata filled with Etag and content length
+     * @throws IOException case etag can't be extracted by s3 or file status
+     */
+    ObjectMetadata getObjectMetadata(URI physicalUri) throws IOException {
+        String bucket = physicalUri.getHost();
+        String key = physicalUri.getPath().substring(1);
+
+        // use underlying filesystem to get the file status and extract
+        // content length and etag (using reflection)
+        Path physicalPath = new Path(physicalUri.getPath());
+        FileStatus fileStatus = this.fs.getFileStatus(physicalPath);
+        try {
+            Method getETagMethod = fileStatus.getClass().getMethod("getETag");
+            String etag = (String) getETagMethod.invoke(fileStatus);
+            // return the two properties over object metadata for easy fallback
+            ObjectMetadata o = new ObjectMetadata();
+            o.setContentLength(fileStatus.getLen());
+            o.setHeader("ETag", etag);
+            return o;
+        } catch (InvocationTargetException | IllegalAccessException | NoSuchMethodException e) {
+            LOG.trace("failed to get etag from file status", e);
```

Comment:
Do we _ever_ expect this to occur?  Basically I don't see this occurring, so I'd like a warning here.  But if I'm missing a case, then sure...

Reply:
sure, this will happen all the time in case if old spark = NoSuchMethodException.
I can put the specific one in ignoring and leave the trace in warn level, it will be a better way to handle this one.